modules/client-handler/src/main/java/org/apache/ignite/client/handler/ClientInboundMessageHandler.java

```diff
@@ -690,21 +689,23 @@ private void processOperation(ChannelHandlerContext ctx, ClientMessageUnpacker i
     }
 
     private void writeFlags(ClientMessagePacker out, ChannelHandlerContext ctx) {
-        boolean assignmentChanged = partitionAssignmentChanged.compareAndSet(true, false);
-
-        if (assignmentChanged && LOG.isInfoEnabled()) {
-            LOG.info("Partition assignment changed, notifying client [connectionId=" + connectionId + ", remoteAddress="
+        // Notify the client about primary replica change that happened for ANY table since the last request.
+        // We can't assume that the client only uses uses a particular table (e.g. the one present in the replica tracker), because
+        // the client can be connected to multiple nodes.
+        long localUpdateCount = primaryReplicaUpdateCount.get();
+        long updateCount = primaryReplicaTracker.updateCount();
+        boolean primaryReplicasChanged = localUpdateCount < updateCount
+                && primaryReplicaUpdateCount.compareAndSet(localUpdateCount, updateCount);
+
+        if (primaryReplicasChanged && LOG.isInfoEnabled()) {
+            LOG.info("Partition primary replica changed, notifying client [connectionId=" + connectionId + ", remoteAddress="
                     + ctx.channel().remoteAddress() + ']');
         }
 
-        var flags = ResponseFlags.getFlags(assignmentChanged);
+        var flags = ResponseFlags.getFlags(primaryReplicasChanged);
```

Comment:
Why not to send primaryReplicaUpdateCount backwards in order to prevent schema notifications reordering?

Reply:
`primaryReplicaUpdateCount` values from different nodes are not comparable.

It's a local count, just a trick to track whether a particular client connection (represented by `ClientInboundMessageHandler`)  has sent the notification flag for the recent update in the (global, one per node) `ClientPrimaryReplicaTracker`.

Reply:
Ok, in that case, like we've discussed in person it's required to use some linearisable "counter" e.g. timestamp as you've proposed. And probably such counter should go to the client and backwards in order to handle races within multi-connections.

Reply:
Done - now we send `maxStartTime` timestamp to the client and back:
* Client can avoid stale updates from a lagging node
* Server will wait for a given timestamp if it lags behind